### ISSUE TITLE
ClimaCore 0.14.13 slows down some GPU simulations and degrades scaling

### ISSUE THREAD

**Issue by Sbozzolo** [CliMA/ClimaCore.jl#1993]
Coupled diagnostic EDMF with 16 horizontal elements (config: config/longrun_configs/amip_target_diagedmf.yml)

[ClimaCore 0.14.13 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-amip/builds/20#0191ddae-d726-4d55-b4f1-7fa91bbcbd90): SYPD ~0.86

[ClimaCore 0.14.12 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-amip/builds/22#0191de1c-b80d-4b65-acbe-7ebd36662366): SYPD ~0.96

[ClimaCore 0.14.13 with ClimaAtmos 0.27.4](https://buildkite.com/clima/climacoupler-amip/builds/24#0191e1b4-d42f-49e8-b0ef-ee63b53aaff9): SYPD ~0.91

[ClimaCore 0.14.13 with ClimaAtmos 0.27.5 but no aerosol](https://buildkite.com/clima/climacoupler-amip/builds/25#0191e213-20e9-4ff0-bb5c-746c4ebb885b): SYPD ~0.92

4 GPUs atmos only EDMF with 30 horizontal elements (config config/benchmark_configs/climaatmos_diagedmf.yml in coupler):

[ClimaCore 0.14.12 with ClimaAtmos 0.27.4](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/185#0191b9f6-b463-4e3c-9d27-8841b0cad56d): SYPD ~1.06

[ClimaCore 0.14.13 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/187#0191e237-c29f-444e-b667-35d517950c89): SYPD ~0.88

[Degradation in scaling in ClimaCore 0.14.13](https://buildkite.com/clima/climacoupler-longruns/builds/810) (the job has two builds one with 1 GPU, the other with 4. SYPD with 4 is only 25 % more) [used to be more like 50%](https://buildkite.com/clima/climacoupler-longruns/builds/800#01918857-262e-4c51-b5fb-21644bdd8ff2)

**Comment by charleskawczynski** [comment]
I think we need to narrow the scope of this issue. First, can we remove the items whose simulations are broken? Relying on progress info during a simulation is not precise.

**Comment by charleskawczynski** [comment]
Doing that, leaves us with

[ClimaCore 0.14.12 with ClimaAtmos 0.27.4](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/185#0191b9f6-b463-4e3c-9d27-8841b0cad56d): SYPD ~1.06

[ClimaCore 0.14.13 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/187#0191e237-c29f-444e-b667-35d517950c89): SYPD ~0.88

Which updates two things at once. I'm not sure that we have good measurements from this to fairly compare.

**Comment by charleskawczynski** [comment]
The builds I ran before merging each change were:

 - https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/314
 - https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/335
 - https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/315

It is peculiar, though. The results do show some variation per function call. In particular, the build where the dss was refactored was significantly better, but then less so after it was merged.

**Comment by Sbozzolo** [comment]
> I think we need to narrow the scope of this issue. First, can we remove the items whose simulations are broken/canceled? Relying on progress info during a simulation is not precise.

There's a 28 minute difference in walltime to reach simulation_time = "13 weeks, 34 minutes" between the coupled run with ClimaCore 0.14.2 and ClimaCore 0.14.3. Even if the simulation was later canceld or failed, that's still evidence of a slowdown. (That's ~7% difference, which is consistent with what we see in SYPD). 

In any case, [this build](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/189#0191fc55-e75b-40f5-b9f0-4c6d4f31212d) is with ClimaCore 0.14.2 and Atmos 0.27.5, 30 elements, runs to completion, and has SYPD of 1.05.

**Comment by charleskawczynski** [comment]
> Even if the simulation was later canceled or failed, that's still evidence of a slowdown. (That's ~7% difference, which is consistent with what we see in SYPD).

Yes, but those measurements are not precise. For example, comparing the last two links at the first step shows the opposite conclusion:

[ClimaCore 0.14.12 with ClimaAtmos 0.27.4](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/185#0191b9f6-b463-4e3c-9d27-8841b0cad56d):  sypd = 0.272

[ClimaCore 0.14.13 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/187#0191e237-c29f-444e-b667-35d517950c89): sypd = 0.369

> In any case, [this build](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/189#0191fc55-e75b-40f5-b9f0-4c6d4f31212d) is with ClimaCore 0.14.2 and Atmos 0.27.5, 30 elements, runs to completion, and has SYPD of 1.05.

Thank you for adding this build, this looks like a good and fair comparison. My best guess is that the launch configuration is not ideal with 4 gpus, since the gpus are probably not fed enough data. This aligns with the observation we saw with lower resolution. It's unfortunate. Perhaps we can try https://github.com/maleadt/StaticCartesian.jl and we can revert to using a linear launch configuration. There is a chance that the issue is with DSS, but I'd be surprised if that were the case since (IIRC) that didn't change the launch configuration.

**Comment by Sbozzolo** [comment]
> > Even if the simulation was later canceled or failed, that's still evidence of a slowdown. (That's ~7% difference, which is consistent with what we see in SYPD).
> 
> Yes, but those measurements are not precise. For example, comparing the last two links at the first step shows the opposite conclusion:
> 
> [ClimaCore 0.14.12 with ClimaAtmos 0.27.4](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/185#0191b9f6-b463-4e3c-9d27-8841b0cad56d): sypd = 0.272
> 
> [ClimaCore 0.14.13 with ClimaAtmos 0.27.5](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/187#0191e237-c29f-444e-b667-35d517950c89): sypd = 0.369

Yes, of course, the first few steps are not significant and they are not representative (we are not even going through the callbacks). But when we are 13 weeks into the simulation, the statistical noise is greatly reduced and I think that can be trusted as a predictor for which build is faster. 

> > In any case, [this build](https://buildkite.com/clima/climacoupler-cpu-gpu-benchmarks/builds/189#0191fc55-e75b-40f5-b9f0-4c6d4f31212d) is with ClimaCore 0.14.2 and Atmos 0.27.5, 30 elements, runs to completion, and has SYPD of 1.05.
> 
> Thank you for adding this build, this looks like a good and fair comparison. My best guess is that the launch configuration is not ideal with 4 gpus, since the gpus are probably not fed enough data. This aligns with the observation we saw with lower resolution. It's unfortunate. Perhaps we can try https://github.com/maleadt/StaticCartesian.jl and we can revert to using a linear launch configuration. There is a chance that the issue is with DSS, but I'd be surprised if that were the case since (IIRC) that didn't change the launch configuration.

Some of the configurations people care about got slower and it would be good to determine what is the root cause and understand the tradeoffs. If the issue is with the benchmarks, let's fix them so that they are reflective of what we care about. 

Our science requirement is to be able to run a multi-year AMIP simulation with a reasonably fast turnaround time.

**Comment by charleskawczynski** [comment]
Actually, I posted the wrong builds that I mentioned that I looked at before merging (cc @szy21). The correct before and after builds are these two:

 - https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/314 (before)
 - https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/339 (after)

Diagnostic edmf went from 977.884 ms to 827.044 ms per step at our target resolution with 1 gpu. Moist held suarez on 4 gpus didn't run in the older build, but if we compare with https://buildkite.com/clima/climaatmos-target-gpu-simulations/builds/336#0191be1d-8c48-4643-8949-5fa0a0c72df0, it also improved. So, I'm now wondering if this issue is coupler-specific.

**Comment by charleskawczynski** [comment]
There is still some sort of regression per the [mentioned build](https://github.com/CliMA/ClimaCore.jl/issues/1993#issuecomment-2353814927), in that the coupler is now slower, it could be that kernels slowed down that are not exercised in ClimaCore/ClimaAtmos.

**Comment by szy21** [comment]
> There is still some sort of regression per the [mentioned build](https://github.com/CliMA/ClimaCore.jl/issues/1993#issuecomment-2353814927), in that the coupler is now slower, it could be that kernels slowed down that are not exercised in ClimaCore/ClimaAtmos.

The slowdown is not only in the coupler, it also shows up in the climaatmos longrun.
ClimaAtmos with old ClimaCore: https://buildkite.com/clima/climaatmos-gpulongruns/builds/387#0191ca6d-b452-426a-8a78-94fcfc841711
ClimaAtmos with new ClimaCore: https://buildkite.com/clima/climaatmos-gpulongruns/builds/389#0191e200-dd62-4fa0-bcb3-559d8e87ddca
There is ~10% slowdown in the aquaplanet with diagnostic EDMF longrun. This is with h_elem=16 though.

**Comment by charleskawczynski** [comment]
Ok, that's helpful. So, it seems like some jobs got faster, others slowed down. Maybe the change in the launch configuration resulted in some speedups and some slowdowns.

**Comment by charleskawczynski** [comment]
(strong scaling, due to resolution changes)

**Comment by charleskawczynski** [comment]
I don’t think that this is actionable, unfortunately. We need more micro benchmarks to catch regressions. Closing as unplanned.
